TRACING.md

```diff
@@ -0,0 +1,312 @@
+sandbox is "destroyed" by the agent shutting down the VM.
+
+Since tracing should work in all scenarios, it cannot be triggered in the
+obvious location of `DestroySandbox`. Therefore, the approach taken is for the
+agent to trigger the end of tracing when the primary workload process exits by
+making `WaitProcess` trigger the gRPC server to end.
+
+# Guest to Host communication using VSOCK
+
+Since [Jaeger does not yet support VSOCK](https://github.com/jaegertracing/jaeger/issues/1019),
+it is necessary to provide a way for the Jaeger client code (running inside the `kata-agent`
+in the VM), to connect to the Jaeger agent (running on the host). The default
+Jaeger port for such communications is
+[UDP port `6831`](https://www.jaegertracing.io/docs/getting-started/).
+
+The required redirection is a achieved using two instances of the `socat(1)` network utility:
+
+- The first instance runs in the guest environment in the VM and forwards all
+  Jaeger trace traffic out of the VM using a VSOCK socket.
+
+- The second instance runs in the host environment and forwards all Jaeger
+  traffic from the VSOCK socket to the Jaeger agent's UDP port.
+
+Further details:
+
+| Name | Type | Environment | Summary | Details |
+|-|-|-|-|-|
+| `jaeger-client-socat-redirector.service` | systemd service | Guest | Guest `UDP:6831` to guest `VSOCK:6831` | Redirects data sent to UDP port `6831` to the host context `VSOCK` socket on port `6831`. |
+| `vsock-to-udp-server.sh` | shell script | Host | Host `VSOCK:6831` to host `UDP:6831` | Redirects all host context VSOCK data on port `6831` to the Jaeger agent, which is listening on UDP port `6831`. |
+
+> **Note:**
+>
+> This is a temporary solution until Jaeger supports VSOCK natively.
+
+For tracing to work, the host system must have a Jaeger agent running.
+
+# Enabling tracing
+
+Tracing is enabled by adding the `agent.trace` option in the `kernel_params=`
+variable in the runtime's `configuration.toml` file.
+
+## Tracing options
+
+Tracing support in the agent is either disabled, or set to one of the
+following:
+
+| Trace type | Trace setting | Description | Use-case | Notes |
+|-|-|-|-|
+| "isolated" | `agent.trace=isolated` (or just `agent.trace`) | The traces only apply to the agent; after the container has been destroyed, the first span will start at agent startup and the last at agent shutdown | Observing agent lifespan. | |
+| "collated" | `agent.trace=collated` | In this mode, spans are associated with their `kata-runtime` initiated counterparts. | Understanding how the runtime calls the agent. | Requires runtime tracing to be enabled in `configuration.toml` (`enable_tracing=true`). |
+
+For example, to enable isolated tracing explicitly, add the following to the
+runtime's `configuration.toml` file:
+
+```
+kernel_params = "agent.trace=isolated ... "
+```
+
+> **Note:**
+>
+> Agent tracing is separate from the tracing for other Kata Containers
```

Comment:
I have a curiosity question - I presume you have traced the agent alongside all the other components, and all the trace items turn up in the right order in the view? I ask as sometimes we get asked about time skew in the VM, and if that existed then maybe it would mess up the trace (presuming Jaeger uses the timestamps, and not the order the messages turn up at the server to sort the events...)

Reply:
afaik, the tracing is correlated in terms of timestamps and span contexts. I haven't seen any issues with "collated tracing mode" (which I think is what you're implicitly referring to), but I'd appreciate it if as many folk as possible could test this PR to further validate ;)